mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala

```diff
@@ -0,0 +1,560 @@
+
+              addTokenToBuffer(
+                wkStartPos,
+                charSize,
+                tokenType,
+                tokeModifier
+              )
+              wkStartPos = wkCurrentPos
+            }
+
+            // Count such as Token.LF
+            if (wkStr == "\n") {
+              currentLine += 1
+              lastNewlineOffset = wkCurrentPos
+            }
+
+          }
+
+        case _ =>
+          val (tokenType, tokeModifier, wkLog) = getTypeAndMod(tk)
+          logString += wkLog
+          addTokenToBuffer(
+            tk.pos.start,
+            tk.text.size,
+            tokenType,
+            tokeModifier
+          )
+      } // end match
+
+    } // end for
+
+    this.logger.info(logString)
+
+    buffer.toList.asJava
+
+  }
+
+  /**
+   * This function returns -1 when capable Type is nothing.
+   *  TokenTypes that can be on multilines are handled in another func.
+   *  See Token.Comment in this file.
+   */
+  private def typeOfNonIdentToken(
+      tk: scala.meta.tokens.Token
+  ): Integer = {
+    tk match {
+      // case _: Token.Ident => // in case of Ident is
+
+      // Alphanumeric keywords
+      case _: Token.ModifierKeyword => getTypeId(SemanticTokenTypes.Modifier)
+      case _: Token.Keyword => getTypeId(SemanticTokenTypes.Keyword)
+
+      // extends Symbolic keywords
+      case _: Token.Hash => getTypeId(SemanticTokenTypes.Keyword)
+      // case _: Token.Colon => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.Viewbound => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.LeftArrow => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.Subtype => getTypeId(SemanticTokenTypes.Keyword)
+      // case _: Token.Equals => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.RightArrow => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.Supertype => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.At => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.Underscore => getTypeId(SemanticTokenTypes.Variable)
+      case _: Token.TypeLambdaArrow => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.ContextArrow => getTypeId(SemanticTokenTypes.Operator)
+
+      // Constant
+      case _: Token.Constant.Int | _: Token.Constant.Long |
+          _: Token.Constant.Float | _: Token.Constant.Double =>
+        getTypeId(SemanticTokenTypes.Number)
+      case _: Token.Constant.Char => getTypeId(SemanticTokenTypes.String)
+
+      // Default
+      case _ => -1
+    }
+
+  }
+
+  class NodeInfo(
+      var tree: cp.Tree,
+      var symbol: cp.Symbol,
+      var pos: scala.reflect.internal.util.Position
+  ) {
+    def this(tree: cp.Tree, pos: cp.Position) = {
+      this(tree, tree.symbol, pos)
+    }
+    def this(imp: cp.Import) = {
+      this(imp, null, null)
+    }
+  }
+  object NodeInfo {
+    def apply[T <: cp.Symbol](sym: T): NodeInfo = {
+      new NodeInfo(null, sym, null)
+    }
+  }
```

Comment:
Also, we could do a bit different structure, which would make some of it a bit simpler:
```
 case class TokenInfo(
      tokenType: String,
      tokenModifiers: List[String]
      pos: Position
  )
```
and we could construct everything in the apply using `getTypeAndMod` method

This would enable us in some cases to do less calculations. For example if we are at